File: packages/js/testkit/src/swap-provider.ts
import { unixSeconds } from "@openreceive/core";
import {
  type OpenReceiveSwapAddressNetwork,
  openReceiveSwapAddressNetworkForPayInAsset,
} from "@openreceive/core/swap-address";
import {
  OPENRECEIVE_SWAP_PAY_IN_ASSETS,
  type SwapAttentionReason,
  type SwapOrder,
  type SwapPayInAsset,
  type SwapProvider,
  type SwapProviderAsset,
  type SwapProviderState,
  type SwapQuote,
} from "@openreceive/node";

/**
 * An in-memory, fully scriptable {@link SwapProvider} for building and
 * testing automated-swap flows offline — no live FixedFloat keys and no real crypto.
 * Drive an attempt through its lifecycle with {@link TestkitSwapProvider.script}, or
 * jump straight to an edge case with `forceRefundRequired`, `forceAttention`, and
 * `forceCreateError`.
 *
 * ```ts
 * const swap = createTestkitSwapProvider();
 * const or = await createOpenReceive({
 *   client: createTestkitReceiveClient(),
 *   swap: { providers: [swap] },
 * });
 * swap.script("USDT_TRON", ["awaiting_deposit", "confirming", "exchanging", "completed"]);
 * // each getStatus poll advances one step, then holds on the last state
 * ```
 */

const NETWORK_DEPOSIT_ADDRESS: Readonly<Record<OpenReceiveSwapAddressNetwork, string>> = {
  TRX: "T9yD14Nj9j7xAB4dbGeiX9h8unkKHxuWwb",
  SOL: "So11111111111111111111111111111111111111112",
  ETH: "0x1111111111111111111111111111111111111111",
};

function depositAddressFor(payInAsset: SwapPayInAsset): string {
  const network = openReceiveSwapAddressNetworkForPayInAsset(payInAsset);
  if (network === undefined) {
    throw new Error(`testkit swap provider has no deposit address for ${payInAsset}`);
  }
  return NETWORK_DEPOSIT_ADDRESS[network];
}

export interface TestkitSwapProviderOptions {
  /** Provider id/name. Must match the provider you register with; defaults to "fixedfloat". */
  readonly name?: string;
  /** Pay-in assets this provider supports. Defaults to the full built-in catalog. */
  readonly supportedAssets?: readonly SwapPayInAsset[];
  /**
   * Clock in unix seconds. Defaults to the real clock, matching
   * TestkitReceiveClient: a fixed low value would mint every swap with a
   * provider_expires_at already in the past. Pass one for deterministic
   * timestamps.
   */
  readonly now?: () => number;
  /** Shadow-invoice expiry the provider requests, in seconds. Defaults to 1800. */
  readonly invoiceExpirySeconds?: number;
  /** Deposit-window length used for provider_expires_at, in seconds. Defaults to 900. */
  readonly depositExpirySeconds?: number;
  /** Per-asset deposit amount the payer must send. Falls back to "1.05". */
  readonly payAmounts?: Partial<Record<SwapPayInAsset, string>>;
}

/** Selector for scripting/forcing a swap: an asset string, or a specific stored attempt. */
export type TestkitSwapSelector =
  | SwapPayInAsset
  | { readonly payInAsset?: SwapPayInAsset; readonly providerOrderId?: string };

interface StoredSwap {
  order: SwapOrder;
  steps: SwapProviderState[];
  next: number;
  attentionReason?: SwapAttentionReason;
}

interface PendingSwapScript {
  readonly steps: readonly SwapProviderState[];
  readonly attentionReason?: SwapAttentionReason;
}

export class TestkitSwapProvider implements SwapProvider {
  readonly name: string;

  createCalls = 0;
  quoteCalls = 0;
  catalogCalls = 0;
  statusCalls = 0;
  readonly createSwapInputs: unknown[] = [];
  readonly quoteInputs: unknown[] = [];
  readonly refundCalls: { readonly providerOrderId: string; readonly refundAddress: string }[] = [];

  readonly #supported: Set<SwapPayInAsset>;
  readonly #now: () => number;
  readonly #invoiceExpirySeconds: number;
  readonly #depositExpirySeconds: number;
  readonly #payAmounts: Partial<Record<SwapPayInAsset, string>>;
  readonly #orders = new Map<string, StoredSwap>();
  readonly #pendingScripts = new Map<SwapPayInAsset, PendingSwapScript>();
  #nextCreateError: Error | undefined;

  constructor(options: TestkitSwapProviderOptions = {}) {
    this.name = options.name ?? "fixedfloat";
    this.#supported = new Set(options.supportedAssets ?? OPENRECEIVE_SWAP_PAY_IN_ASSETS);
    this.#now = options.now ?? unixSeconds;
    this.#invoiceExpirySeconds = options.invoiceExpirySeconds ?? 1800;
    this.#depositExpirySeconds = options.depositExpirySeconds ?? 900;
    this.#payAmounts = options.payAmounts ?? {};
  }

  async supportedPayInAssets(): Promise<Set<SwapPayInAsset>> {
    return new Set(this.#supported);
  }

  async payInAssetCatalog(): Promise<readonly SwapProviderAsset[]> {
    this.catalogCalls += 1;
    return Array.from(this.#supported, (payInAsset) => ({
      pay_asset: payInAsset,
      available: true,
      minimum_pay_amount: "1",
      maximum_pay_amount: "5000",
    }));
  }

  invoiceExpirySeconds(): number {
    return this.#invoiceExpirySeconds;
  }

  async quote(input: {
    readonly payInAsset: SwapPayInAsset;
    readonly invoiceAmountMsats: number;
  }): Promise<SwapQuote> {
    this.quoteCalls += 1;
    this.quoteInputs.push(input);
    return {
      pay_amount: this.#payAmountFor(input.payInAsset),
      pay_asset: input.payInAsset,
      available: true,
      provider: this.name,
      minimum_pay_amount: "1",
      maximum_pay_amount: "5000",
    };
  }

  async createSwap(input: {
    readonly payInAsset: SwapPayInAsset;
    readonly bolt11: string;
    readonly invoiceAmountMsats: number;
  }): Promise<SwapOrder> {
    this.createCalls += 1;
    this.createSwapInputs.push(input);
    if (this.#nextCreateError !== undefined) {
      const error = this.#nextCreateError;
      this.#nextCreateError = undefined;
      throw error;
    }

    const providerOrderId = `testkit-swap-${this.createCalls}`;
    const order: SwapOrder = {
      provider: this.name,
      provider_order_id: providerOrderId,
      provider_token: `testkit-token-${this.createCalls}`,
      pay_in_asset: input.payInAsset,
      deposit_address: depositAddressFor(input.payInAsset),
      deposit_amount: this.#payAmountFor(input.payInAsset),
      expires_at: this.#now() + this.#depositExpirySeconds,
      state: "awaiting_deposit",
    };
    const pending = this.#pendingScripts.get(input.payInAsset);
    this.#orders.set(providerOrderId, {
      order,
      steps: pending === undefined ? [] : [...pending.steps],
      next: 0,
      ...(pending?.attentionReason === undefined
        ? {}
        : { attentionReason: pending.attentionReason }),
    });
    return order;
  }

  async getStatus(order: SwapOrder): Promise<SwapOrder> {
    this.statusCalls += 1;
    const stored = this.#orders.get(order.provider_order_id);
    if (stored === undefined) return order;
    if (stored.next < stored.steps.length) {
      const nextState = stored.steps[stored.next];
      stored.next += 1;
      stored.order = applyState(stored.order, nextState);
      if (nextState === "attention" && stored.attentionReason !== undefined) {
        stored.order = { ...stored.order, attention_reason: stored.attentionReason };
      }
    }
    return stored.order;
  }

  async requestRefund(order: SwapOrder, refundAddress: string): Promise<void> {
    this.refundCalls.push({ providerOrderId: order.provider_order_id, refundAddress });
    const stored = this.#orders.get(order.provider_order_id);
    if (stored !== undefined) {
      stored.order = applyState(stored.order, "refund_pending");
    }
  }

  /**
   * Queue a sequence of provider states. Each `getStatus` poll advances one step and
   * then holds on the last state. Scripting an asset before a swap is started attaches
   * the sequence to the next attempt created for that asset.
   */
  script(selector: TestkitSwapSelector, states: readonly SwapProviderState[]): void {
    if (states.length === 0) {
      throw new RangeError("swap script must include at least one state");
    }
    const matched = this.#match(selector);
    for (const stored of matched) {
      stored.steps = [...states];
      stored.next = 0;
      delete stored.attentionReason;
    }
    const payInAsset = selectorAsset(selector);
    if (payInAsset !== undefined) {
      this.#pendingScripts.set(payInAsset, { steps: [...states] });
    }
  }

  /**
   * Force `refund_required`: existing selected attempts report it on their next
   * `getStatus`, and like {@link script} an asset selector also queues it for
   * attempts created later.
   */
  forceRefundRequired(selector: TestkitSwapSelector): void {
    this.#force(selector, "refund_required");
  }

  /**
   * Force the `attention` state with a recorded reason: existing selected
   * attempts move immediately, and like {@link script} an asset selector also
   * queues it for attempts created later.
   */
  forceAttention(
    selector: TestkitSwapSelector,
    reason: SwapAttentionReason = "provider_reported_emergency",
  ): void {
    this.#force(selector, "attention", reason);
  }

  /** Make the next `createSwap` call reject, simulating a provider order-creation failure. */
  forceCreateError(error: Error = new Error("testkit swap provider create failed")): void {
    this.#nextCreateError = error;
  }

  #force(
    selector: TestkitSwapSelector,
    state: SwapProviderState,
    attentionReason?: SwapAttentionReason,
  ): void {
    for (const stored of this.#match(selector)) {
      stored.steps = [];
      stored.next = 0;
      delete stored.attentionReason;
      stored.order = applyState(stored.order, state);
      if (attentionReason !== undefined) {
        stored.order = { ...stored.order, attention_reason: attentionReason };
      }
    }
    const payInAsset = selectorAsset(selector);
    if (payInAsset !== undefined) {
      this.#pendingScripts.set(payInAsset, {
        steps: [state],
        ...(attentionReason === undefined ? {} : { attentionReason }),
      });
    }
  }

  #match(selector: TestkitSwapSelector): StoredSwap[] {
    const asset = selectorAsset(selector);
    const providerOrderId = typeof selector === "object" ? selector.providerOrderId : undefined;
    return [...this.#orders.values()].filter((stored) => {
      if (providerOrderId !== undefined && stored.order.provider_order_id !== providerOrderId) {
        return false;
      }
      if (asset !== undefined && stored.order.pay_in_asset !== asset) return false;
      return true;
    });
  }

  #payAmountFor(payInAsset: SwapPayInAsset): string {
    return this.#payAmounts[payInAsset] ?? "1.05";
  }
}

export function createTestkitSwapProvider(
  options: TestkitSwapProviderOptions = {},
): TestkitSwapProvider {
  return new TestkitSwapProvider(options);
}

function selectorAsset(selector: TestkitSwapSelector): SwapPayInAsset | undefined {
  return typeof selector === "string" ? selector : selector.payInAsset;
}

function applyState(order: SwapOrder, state: SwapProviderState): SwapOrder {
  const attention = state === "attention";
  return {
    ...order,
    state,
    ...(atOrAfter(state, "confirming") ? { deposit_tx_id: "testkit-deposit-tx" } : {}),
    ...(state === "completed" ? { payout_tx_id: "testkit-payout-tx" } : {}),
    ...(state === "refunded" ? { refund_tx_id: "testkit-refund-tx" } : {}),
    ...(attention ? { attention: true } : {}),
  };
}

const PROGRESS_ORDER: readonly SwapProviderState[] = [
  "creating_provider_order",
  "awaiting_deposit",
  "confirming",
  "exchanging",
  "paying_invoice",
  "completed",
];

/** True once the payer's deposit has been detected (confirming or later progress state). */
function atOrAfter(state: SwapProviderState, floor: SwapProviderState): boolean {
  const stateIndex = PROGRESS_ORDER.indexOf(state);
  const floorIndex = PROGRESS_ORDER.indexOf(floor);
  return stateIndex >= 0 && floorIndex >= 0 && stateIndex >= floorIndex;
}

File: examples/buttons/server/rails/lib/button_shop/testkit/swap_provider.rb
# frozen_string_literal: true

module ButtonShop
  module Testkit
    # An in-memory swap provider, and a PORT of packages/js/testkit's
    # TestkitSwapProvider — same deposit addresses, same `testkit-swap-N` order
    # ids, same "1.05" pay amount — so the same Playwright assertions hold
    # against Rails and against the Node stacks.
    #
    # The contract is duck-typed by OpenReceive::Server::Service, and it is the
    # FixedFloat provider's public surface: `name`, `supported_pay_in_assets`,
    # `pay_in_asset_catalog`, `invoice_expiry_seconds`, `quote`, `create_swap`,
    # `get_status`, `request_refund`. The runtime attachments the real provider
    # takes (`attach_swap_cache` and friends) are all `respond_to?`-guarded by
    # the service, so a fake that has none of them is simply left alone.
    class SwapProvider
      # One address per NETWORK, not per asset: USDT_TRON and a hypothetical
      # second Tron token share a Tron address, which is exactly the ambiguity
      # the deposit panel's warning is about.
      NETWORK_DEPOSIT_ADDRESS = {
        "TRX" => "T9yD14Nj9j7xAB4dbGeiX9h8unkKHxuWwb",
        "SOL" => "So11111111111111111111111111111111111111112",
        "ETH" => "0x1111111111111111111111111111111111111111"
      }.freeze

      PAY_AMOUNT = "1.05"
      # The shadow invoice must outlive the provider order; the service takes
      # this as a FLOOR when it mints.
      INVOICE_EXPIRY_SECONDS = 1_800
      DEPOSIT_EXPIRY_SECONDS = 900

      PROGRESS_ORDER = %w[
        creating_provider_order awaiting_deposit confirming exchanging paying_invoice completed
      ].freeze

      attr_reader :name

      def initialize(name: "fixedfloat", clock: -> { Time.now.to_i })
        @name = name
        @clock = clock
        @orders = {}
        # An asset scripted BEFORE any attempt exists arms the next attempt for
        # it — the harness advances a swap it has not created yet.
        @pending = {}
        @create_calls = 0
        @quote_calls = 0
        @status_calls = 0
        @refund_calls = []
        @lock = Mutex.new
      end

      def supported_pay_in_assets
        OpenReceive::Server::Swap::Assets::PAY_IN_ASSETS.dup
      end

      def pay_in_asset_catalog
        supported_pay_in_assets.map do |pay_in_asset|
          {
            "pay_asset" => pay_in_asset,
            "available" => true,
            "minimum_pay_amount" => "1",
            "maximum_pay_amount" => "5000"
          }
        end
      end

      def invoice_expiry_seconds(pay_in_asset: nil)
        INVOICE_EXPIRY_SECONDS
      end

      def quote(pay_in_asset:, invoice_amount_msats:)
        @lock.synchronize { @quote_calls += 1 }
        {
          "pay_amount" => PAY_AMOUNT,
          "pay_asset" => pay_in_asset,
          "available" => true,
          "provider" => @name,
          "minimum_pay_amount" => "1",
          "maximum_pay_amount" => "5000"
        }
      end

      def create_swap(pay_in_asset:, bolt11:, invoice_amount_msats:)
        @lock.synchronize do
          @create_calls += 1
          provider_order_id = "testkit-swap-#{@create_calls}"
          order = {
            "provider" => @name,
            "provider_order_id" => provider_order_id,
            "provider_token" => "testkit-token-#{@create_calls}",
            "pay_in_asset" => pay_in_asset,
            "deposit_address" => deposit_address_for(pay_in_asset),
            "deposit_amount" => PAY_AMOUNT,
            "expires_at" => @clock.call + DEPOSIT_EXPIRY_SECONDS,
            "state" => "awaiting_deposit"
          }
          armed = @pending.delete(pay_in_asset)
          @orders[provider_order_id] = {
            "order" => order,
            "steps" => armed.nil? ? [] : armed.fetch("steps").dup,
            "next" => 0,
            "attention_reason" => armed&.dig("attention_reason")
          }
          order
        end
      end

      # One step per poll, then hold on the last state — the harness advances a
      # swap by letting the page poll, which is how a payer experiences it.
      def get_status(order)
        stored_id = OpenReceive.stringify(order).fetch("provider_order_id")
        @lock.synchronize do
          @status_calls += 1
          entry = @orders[stored_id]
          return order if entry.nil?

          if entry.fetch("next") < entry.fetch("steps").length
            state = entry.fetch("steps")[entry.fetch("next")]
            entry["next"] += 1
            entry["order"] = apply_state(entry.fetch("order"), state, entry["attention_reason"])
          end
          entry.fetch("order")
        end
      end

      def request_refund(order, refund_address)
        stored_id = OpenReceive.stringify(order).fetch("provider_order_id")
        @lock.synchronize do
          @refund_calls << { "provider_order_id" => stored_id, "refund_address" => refund_address }
          entry = @orders[stored_id]
          entry["order"] = apply_state(entry.fetch("order"), "refund_pending", nil) unless entry.nil?
        end
        nil
      end

      # ------------------------------------------------------------- controls

      # Queue states for the selected attempts, and arm the asset so an attempt
      # created later gets them too.
      def script(selector, states, attention_reason: nil)
        raise ArgumentError, "swap script must include at least one state" if states.empty?

        @lock.synchronize do
          match(selector).each do |entry|
            entry["steps"] = states.dup
            entry["next"] = 0
            entry["attention_reason"] = attention_reason
          end
          asset = selector["pay_in_asset"]
          @pending[asset] = { "steps" => states.dup, "attention_reason" => attention_reason } if asset
        end
      end

      # `refund_required` and `attention` land IMMEDIATELY rather than on the
      # next poll: they are the states a test jumps to, and waiting a poll for
      # them only makes the harness flakier.
      def force(selector, state, attention_reason: nil)
        @lock.synchronize do
          match(selector).each do |entry|
            entry["steps"] = []
            entry["next"] = 0
            entry["attention_reason"] = attention_reason
            entry["order"] = apply_state(entry.fetch("order"), state, attention_reason)
          end
          asset = selector["pay_in_asset"]
          @pending[asset] = { "steps" => [state], "attention_reason" => attention_reason } if asset
        end
      end

      def counters
        @lock.synchronize do
          {
            "create_calls" => @create_calls,
            "quote_calls" => @quote_calls,
            "status_calls" => @status_calls,
            "refund_calls" => @refund_calls.map(&:dup)
          }
        end
      end

      private

      def match(selector)
        provider_order_id = selector["provider_order_id"]
        asset = selector["pay_in_asset"]
        @orders.values.select do |entry|
          order = entry.fetch("order")
          next false if provider_order_id && order.fetch("provider_order_id") != provider_order_id
          next false if asset && order.fetch("pay_in_asset") != asset

          true
        end
      end

      def apply_state(order, state, attention_reason)
        next_order = order.merge("state" => state)
        next_order["deposit_tx_id"] = "testkit-deposit-tx" if at_or_after?(state, "confirming")
        next_order["payout_tx_id"] = "testkit-payout-tx" if state == "completed"
        next_order["refund_tx_id"] = "testkit-refund-tx" if state == "refunded"
        if state == "attention"
          next_order["attention"] = true
          next_order["attention_reason"] = attention_reason unless attention_reason.nil?
        end
        next_order
      end

      def at_or_after?(state, floor)
        state_index = PROGRESS_ORDER.index(state)
        floor_index = PROGRESS_ORDER.index(floor)
        !state_index.nil? && !floor_index.nil? && state_index >= floor_index
      end

      def deposit_address_for(pay_in_asset)
        network = OpenReceive::Server::Swap::Assets::ASSET_INFO
                  .fetch(pay_in_asset).fetch("network")
        NETWORK_DEPOSIT_ADDRESS.fetch(network)
      end
    end
  end
end
